src/java/org/apache/cassandra/index/sai/disk/v1/trie/InvertedIndexWriter.java

```diff
@@ -0,0 +1,112 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1.trie;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.commons.lang3.mutable.MutableLong;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.TermsIterator;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.disk.v1.postings.PostingsWriter;
+import org.apache.cassandra.index.sai.utils.SAICodecUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+/**
+ * Builds an on-disk inverted index structure: terms dictionary and postings lists.
+ */
+@NotThreadSafe
+public class InvertedIndexWriter implements Closeable
+{
+    private final TrieTermsDictionaryWriter termsDictionaryWriter;
+    private final PostingsWriter postingsWriter;
+    private long postingsAdded;
+
+    public InvertedIndexWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, boolean segmented) throws IOException
+    {
+        this.termsDictionaryWriter = new TrieTermsDictionaryWriter(indexDescriptor, indexContext, segmented);
+        this.postingsWriter = new PostingsWriter(indexDescriptor, indexContext, segmented);
+    }
+
+    /**
+     * Appends a set of terms and associated postings to their respective overall SSTable component files.
+     *
+     * @param terms an iterator of terms with their associated postings
+     *
+     * @return metadata describing the location of this inverted index in the overall SSTable
+     *         terms and postings component files
+     */
+    public SegmentMetadata.ComponentMetadataMap writeAll(TermsIterator terms) throws IOException
+    {
+        // Terms and postings writers are opened in append mode with pointers at the end of their respective files.
+        long termsOffset = termsDictionaryWriter.getStartOffset();
+        long postingsOffset = postingsWriter.getStartOffset();
```

Comment:
...and this matters because segments are written to the same file, correct? The reader might guess that from the segment metadata below, but I wonder if we could clarify that in docs and maybe even call the method `writeSegment()` or something similar.

Reply:
It is not always writing segments. This depends on whether we are doing an initial write (in which case it will be segmented) or whether we are compacting the index (in which case it writes a single segment).

Saying that, it is always writing a segment even it there is only one of them. I'd like to propose `writeCompleteSegment` as the method name as this accurately reflects what it is doing.

WDYT?

Reply:
It's better than `writeAll()` :)

Reply:
...or we could just rename the class `TrieSegmentWriter` and the method `write()` :D

Reply:
I think the naming of this class as `InvertedIndexWriter` was to try and abstract it from the low-level implementation. While I prefer this naming I think that this class doesn't really sit will with the naming used elsewhere. We have generally gone with `literal` and `numeric` as differentiaters between the 2 types of indexes supported.

I'm going to rename the class `LiteralIndexWriter` and keep the method as `writeCompleteSegment`.

Reply:
What about `LiteralIndexSegmentWriter` (or `LiteralSegmentWriter`) and `write()`? Here's my reasoning:

- The class only does one thing...writes segments for literal indexes (I think?)
- I'm not sure what a "complete" segment means if `segmented` is `true` and we open temporary files (...and if we remove segment compaction, everything is "complete").